app/models/password_recovery_form.rb

```diff
@@ -0,0 +1,22 @@
+class PasswordRecoveryForm
+  include ActiveModel::Model
+
+  attr_reader :code, :password, :password_confirmation
+  attr_accessor :email
+
+  validates_presence_of :code, :email, :password, :password_confirmation
+  validates_confirmation_of :password
+  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }
+  validates :password, length: { minimum: 8 }
+
+  def initialize(params)
+    @code = params[:code]
+    @email = params[:email]
+    @password = params[:password]
+    @password_confirmation = params[:password_confirmation]
+  end
+
+  def to_h
+    instance_variables.map { |var| [var.to_s.delete('@').to_sym, instance_variable_get(var)] }.to_h
+  end
```

Comment:
These `to_h` methods are repeated a lot now. We use them for debugging, is that right? In which case to save repeating them all the time maybe put them into a helper for reuse?

Reply:
Ruby says you should define your own to_h method on your classes that use it.  The Authentication backend methods accept hashes rather than objects.